File: ruby/grade-school/grade_school.rb
class School
  def initialize
    @students = {}
  end

  def students(grade)
    @students.fetch(grade, [])
  end

  def add(name, grade)
    students_in_grade = students(grade)
    students_in_grade << name
    @students[grade] = students_in_grade.sort
    true
  end

  def students_by_grade
    @students.map do |grade, students|
      { grade: grade, students: students }
    end.sort_by { |el| el[:grade] }
  end
end

module BookKeeping
  VERSION = 3
end

File: python/grade-school/grade_school.py
class School:
    def __init__(self, school_name):
        self.school_name = school_name
        self.students = {}

    def add(self, name, grade):
        self._students_in_grade(grade).append(name)

    def grade(self, number):
        return set(self._students_in_grade(number))

    def sort(self):
        return [(g, tuple(sorted(s))) for g, s in self.students.items()]

    def _students_in_grade(self, grade):
        if grade not in self.students:
            self.students[grade] = []

        return self.students[grade]
